src/engine/ChunkSystemTileMap.ts

```diff
@@ -0,0 +1,466 @@
+import { Actor } from './Actor';
+import { Vector } from './Algebra';
+import { Class } from './Class';
+import { BoundingBox } from './Collision/BoundingBox';
+import { Configurable } from './Configurable';
+import { Engine } from './Engine';
+import * as Events from './Events';
+import { SpriteSheet } from './Drawing/SpriteSheet';
+import { Cell, TileMap } from './TileMap';
+
+export type ChunkGenerator = (chunkColumn: number, chunkRow: number, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => TileMap;
+export type SimpleChunkGenerator = (
+  chunk: TileMap,
+  chunkCellColumn: number,
+  chunkCellRow: number,
+  chunkSystemTileMap: ChunkSystemTileMap,
+  engine: Engine
+) => TileMap;
+export type SimpleCellGenerator = (
+  cell: Cell,
+  cellColumn: number,
+  cellRow: number,
+  chunk: TileMap,
+  chunkSystemTileMape: ChunkSystemTileMap,
+  engine: Engine
+) => void;
+
+export type ChunkSystemGarbageCollectorPredicate = (chunk: TileMap, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => boolean;
+
+export type ChunkRenderingCachePredicate = (chunk: TileMap, chunkSystemTileMap: ChunkSystemTileMap, engine: Engine) => boolean;
+
+type CachedTileMap = TileMap & { renderingCache: null | HTMLCanvasElement };
+
+interface ChunkSystemTileMapArgs {
+  x: number;
+  y: number;
+  chunkSize: number;
+  cellWidth: number;
+  cellHeight: number;
+  rows: number;
+  cols: number;
+  chunkGenerator: ChunkGenerator;
+  chunkGarbageCollectorPredicate: ChunkSystemGarbageCollectorPredicate;
+  chunkRenderingCachePredicate?: null | ChunkRenderingCachePredicate;
+}
+
+/**
+ * @hidden
+ */
+export class ChunkSystemTileMapImpl extends Class {
+  public readonly x: number;
+  public readonly y: number;
+  public readonly cellWidth: number;
+  public readonly cellHeight: number;
+  public readonly chunkSize: number;
+  public readonly cols: number;
+  public readonly rows: number;
+  public readonly chunkCols: number;
+  public readonly chunkRows: number;
+  public readonly chunkGenerator: ChunkGenerator;
+  public readonly chunkGarbageCollectorPredicate: ChunkSystemGarbageCollectorPredicate;
```

Comment:
It all comes down to the expected maximum size of the game world. For a "common" game, a `() => false` predicate is the best option since the world should fit into the memory. I am willing to make this the default implementation, if you agree.

However, for games aiming for a very large (or seemingly inifnite) world, it would be recommended to return `true` for any chunk that is "far enough" from the viewport. Such game would however need to handle the cases of actors trying to pathfind through garbage-collected chunks - it might be necessary for the predicate to communicate with other parts of the game's logic.

Reply:
After some thought, I've made the parameter optional, defaulting to no garbage collection, which is the best (safest) option for many games.